db/similarity.py

```diff
@@ -0,0 +1,144 @@
+        batch_query = text("""
+            SELECT ll.id
+                 , llj.data AS ll_data
+                 , jsonb_object_agg(hlm.model, hlm.data) as hl_data
+              FROM (
+            SELECT id 
+              FROM lowlevel 
+         LEFT JOIN similarity 
+             USING (id) 
+             WHERE similarity.id is NULL) ll
+              JOIN lowlevel_json AS llj USING (id)
+              JOIN highlevel_model AS hlm USING (id) 
+          GROUP BY (ll.id, llj.data)
+             LIMIT :batch_size
+        """)
+
+        while True:
+            with connection.begin():
+                result = connection.execute(batch_query, {"batch_size": batch_size})
+                if not result.rowcount:
+                    break
+
+                row = result.fetchone()
+                while row:
+                    lowlevel = row["ll_data"]
+                    models = row["hl_data"]
+                    data = (lowlevel, models)
+                    submit_similarity_by_id(row["id"], data=data, metrics=metrics, connection=connection)
+                    row = result.fetchone()
+
+            sim_count = count_similarity()
+            print("Processed {} / {} ({:.3f}%)".format(sim_count,
+                                                       lowlevel_count,
+                                                       float(sim_count) / lowlevel_count * 100))
+
+
+def insert_similarity(connection, id, vectors, metric_names):
+    """Inserts a row of similarity vectors for a given lowlevel.id into
+    the similarity table.
+
+        Args: id: lowlevel.id to be submitted
+              vectors: list of metric vectors for a recording
+              metric_names: corresponding list of metric names
+    """
+    params = {}
+    params["id"] = id
+    for name, vector in zip(metric_names, vectors):
+        params[name] = list(vector)
+
+    query = text("""
+        INSERT INTO similarity (
+                    id, %(names)s)
+             VALUES ( 
+                    :id, %(values)s)
+        ON CONFLICT (id)
+         DO NOTHING
+    """ % {"names": ', '.join(metric_names),
+           "values": ':' + ', :'.join(metric_names)})
+    connection.execute(query, params)
+
+
+def count_similarity():
+    # Get total number of submissions in similarity table
+    with db.engine.connect() as connection:
+        query = text("""
+            SELECT COUNT(*)
+              FROM similarity
+        """)
+        result = connection.execute(query)
+        return result.fetchone()[0]
+
+
+def submit_similarity_by_id(id, data=None, metrics=None, connection=None):
+    """Computes similarity metrics for a single recording specified
+    by lowlevel.id, then inserts the metrics as a new row in the
+    similarity table."""
+    try:
+        id = int(id)
+    except ValueError:
+        raise BadDataException('Parameter `id` must be an integer.')
+
+    if not metrics:
+        metrics = similarity.utils.init_metrics()
+
+    if not data:
```

Comment:
I added this so we can use the same function when we compute metrics for a new recording submitted to the db.